pytng/pytng.pyx

```diff
@@ -93,25 +407,206 @@ cdef class MemoryWrapper:
+        stat = tng_num_frames_get(self._traj, & self._n_frames)
+        if stat != TNG_SUCCESS:
+            raise IOError("Number of frames cannot be read")
+
+        stat = tng_num_particles_get(self._traj, & self._n_particles)
+        if stat != TNG_SUCCESS:
+            raise IOError("Number of particles cannot be read")
+
+        # NOTE can we just loop over this directly in some way?
+        stat = tng_num_frame_sets_get(self._traj, & self._n_frame_sets)
+        # TODO can this be read straight from the struct as self._traj->n_frame_sets?
+        # they note that this is not always updated
+
+        cdef int64_t exponent
+        stat = tng_distance_unit_exponential_get(self._traj, & exponent)
+        if stat != TNG_SUCCESS:
+            raise IOError("Distance exponent cannot be read")
+
+        self._distance_scale = 10.0**(exponent+9)
+
+        self.is_open = True
+        self.step = 0
+        self.reached_eof = False
+    
+
+    def _close(self):
+        """Make sure the file handle is closed"""
+        if self.is_open:
+            tng_util_trajectory_close(& self._traj)
+            self.is_open = False
+            self._n_frames = -1
+
+    def spool(self):
+        cdef tng_function_status stat = TNG_SUCCESS
+        cdef int64_t block_count = 0
+        while stat != TNG_CRITICAL:
+            print("block read called {}".format(block_count))
+            stat = self._read_next_block()
+            block_count += 1
+
+    # NOTE this looks simple and may work ? may be too low level as it all hinges on whether the file positions are kept up to date
+    # can we just fseek to the first TFS?
+    cdef tng_function_status _read_next_block(self):
+        cdef tng_function_status stat
+        cdef tng_gen_block* block
+        cdef int64_t block_id = -1
+        stat = tng_block_init(& block)
+        if stat != TNG_SUCCESS:
+            return TNG_CRITICAL
+        
+        fseeko(dereference(self._traj).input_file, 0, SEEK_SET) # set to start of inut file
+
+        #tng_data->current_trajectory_frame_set_input_file_pos = orig_frame_set_file_pos;?
+
+        stat = tng_block_header_read(self._traj, block)
+        if stat != TNG_SUCCESS:
+            tng_block_destroy(&block)
+            return TNG_CRITICAL
+            
+        stat = tng_block_read_next(self._traj, block, TNG_SKIP_HASH)
+        if stat != TNG_SUCCESS:
+            tng_block_destroy(& block)
+            return TNG_CRITICAL
+        
+        block_id = dereference(block).id
+        printf("\n %ld \n", block_id)
+        cdef name_len = strlen(dereference(block).name) +1
+        cdef bname = <char*> malloc(name_len * sizeof(char)) # TNG_MAX_STR_LEN = 1024
+        bname = dereference(block).name
+        printf(bname)
+        return TNG_SUCCESS
+
+    cdef _block_interpret(self):
+        pass
+        #logic to interpret the block types here
+
+    cdef _block_numpy_cast(self):
```

Comment:
I think the signature for this should be that a numpy array gets passed in, shape checked then the pointer from it passed into TNG.  It's then the responsibility of the `TNGTrajectoryReader` to make appropriate (malloc'ed) ones for pos/vel/force.

Reply:
Noted